v1/src/main/java/com/google/cloud/teleport/templates/common/CsvConverters.java

```diff
@@ -0,0 +1,306 @@
+              CSVFormat format = CSVFormat.Predefined.valueOf(input.getX()).getFormat();
+              return String.valueOf(format.getDelimiter());
+            }
+      });
+    }
+  }
+
+  /**
+   * The {@link ReadCsv} class is a {@link PTransform} that reads from one for more Csv files. The
+   * transform returns a {@link PCollectionTuple} consisting of the following {@link PCollection}:
+   *
+   * <ul>
+   *   <li>{@link ReadCsv#headerTag()} - Contains headers found in files if read with headers,
+   *       contains empty {@link PCollection} if no headers.
+   *   <li>{@link ReadCsv#lineTag()} - Contains Csv lines as a {@link PCollection} of strings.
+   * </ul>
+   */
+  @AutoValue
+  public abstract static class ReadCsv extends PTransform<PBegin, PCollectionTuple> {
+
+    public static Builder newBuilder() {
+      return new AutoValue_CsvConverters_ReadCsv.Builder();
+    }
+
+    public abstract ValueProvider<String> csvFormat();
+
+    @Nullable
+    public abstract ValueProvider<String> delimiter();
+
+    public abstract ValueProvider<Boolean> hasHeaders();
+
+    public abstract ValueProvider<String> inputFileSpec();
+
+    public abstract TupleTag<String> headerTag();
+
+    public abstract TupleTag<String> lineTag();
+
+    public abstract ValueProvider<String> fileEncoding();
+
+    @Override
+    public PCollectionTuple expand(PBegin input) {
+
+      if (hasHeaders().get()) {
+        return input
+            .apply("MatchFilePattern", FileIO.match().filepattern(inputFileSpec()))
+            .apply("ReadMatches", FileIO.readMatches())
+            .apply(
+                "ReadCsvWithHeaders",
+                ParDo.of(
+                        new GetCsvHeadersFn(
+                            headerTag(), lineTag(), csvFormat(), delimiter(), fileEncoding()))
+                    .withOutputTags(headerTag(), TupleTagList.of(lineTag())));
+      }
+
+      return PCollectionTuple.of(
+          lineTag(), input.apply("ReadCsvWithoutHeaders", TextIO.read().from(inputFileSpec())));
+    }
+
+    /** Builder for {@link ReadCsv}. */
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder setCsvFormat(ValueProvider<String> csvFormat);
+
+      public abstract Builder setDelimiter(@Nullable ValueProvider<String> delimiter);
+
+      public abstract Builder setHasHeaders(ValueProvider<Boolean> hasHeaders);
+
+      public abstract Builder setInputFileSpec(ValueProvider<String> inputFileSpec);
+
+      public abstract Builder setHeaderTag(TupleTag<String> headerTag);
+
+      public abstract Builder setLineTag(TupleTag<String> lineTag);
+
+      public abstract Builder setFileEncoding(ValueProvider<String> fileEncoding);
+
+      abstract ReadCsv autoBuild();
+
+      public ReadCsv build() {
+
+        ReadCsv readCsv = autoBuild();
+
+        checkArgument(readCsv.inputFileSpec() != null, "Input file spec must be provided.");
```

Comment:
how to perform the check since it's value provider now?

Reply:
unfortunately, only inside a DoFn (or a flex template)

for classic template the program already ran and the graph is ready, you just inject the value in the placeholders (ValueProviders), so the model is not imperative